src/types/dto-decorator-factories.type.ts

```diff
@@ -0,0 +1,55 @@
+import { DTODecoratorName } from './dto-decorator-name.type';
+import { DTODecoratorOptions } from './dto-decorator-options.type';
+import { PropertyDecoratorFactory } from './property-decorator-factory.type';
+import { Target } from './target.type';
+
+/* Consumers of the library will import (or create) a suite of factories that create property decorators for the
+ * suppored set of decorators and options.
+ */
+export type DTODecoratorFactories = {
+    [Name in DTODecoratorName]: PropertyDecoratorFactory<DTODecoratorOptions<Name>>;
+};
+
+/* Multiple property decorators can be composed into a new property decorator.
+ */
+export function composePropertyDecorators(
+    decorators: PropertyDecorator[],
+): PropertyDecorator {
+    return (target: Target, propertyKey: string | symbol): void => {
+        for (const decorator of decorators) {
+            decorator(target, propertyKey);
+        }
+    };
+}
+
+/* Multiple property decorator factories of the same type can be composed into a new factory.
+ */
+export function composePropertyDecoratorFactories<Name extends DTODecoratorName>(
+    factories: PropertyDecoratorFactory<DTODecoratorOptions<Name>>[],
+): PropertyDecoratorFactory<DTODecoratorOptions<Name>> {
+    return (options: DTODecoratorOptions<Name>): PropertyDecorator => composePropertyDecorators(
+        factories.map(
+            (factory) => factory(options),
+        ),
+    );
+}
+
+/* A specific property decorator factory can be selected from a factory by name.
+ */
+export function selectPropertyDecoratorFactory<Name extends DTODecoratorName>(
+    name: Name,
+    factory: DTODecoratorFactories,
+): PropertyDecoratorFactory<DTODecoratorOptions<Name>> {
+    return factory[name] as unknown as PropertyDecoratorFactory<DTODecoratorOptions<Name>>;
```

Comment:
I could not find any way to avoid the cast here. Even though both `factory` and `DTODecoratorOptions` and  are parameterized on `Name`, `tsc` cannot reason that `factory[name]` uses `DTODecoratorOptions<Name>` and we get compilation errors related to option incompatibility.

I'd really hope there's a solution.